### PR TITLE
16353 Extend TF Pascal VOC exporter

### PR DESCRIPTION
**Pull request by JacopoMangiavacchi**
Align v2 TensorFlow Pascal VOC exporter to all features currently supported on V1

### CODE REVIEW

**Comment by JacopoMangiavacchi** [comment]
@wbreza test are failing again on expect(exportProjectSpy).not.toBeCalled();  I'll better investigate on this and also I'll probably need to implement a couple of more tests.

File: src/providers/export/tensorFlowPascalVOC.test.ts [hunk 1]
```diff
@@ -27,9 +38,12 @@ describe("TFPascalVOC Json Export Provider", () => {
 
     const tagLengthInPbtxt = 37;
 
+    // jest.setTimeout(20000);
```
Comment (wbreza) [inline src/providers/export/tensorFlowPascalVOC.test.ts]: Nit: Remove comments.

File: src/providers/export/tensorFlowPascalVOC.ts [hunk 2]
```diff
@@ -64,71 +85,92 @@ export class TFPascalVOCJsonExportProvider extends ExportProvider<ITFPascalVOCJs
...
+        });
+
+        try {
+            await Promise.all(allImageExports);
+        } catch (err) {
+            console.log(err);
+        }
+    }
+
+    private async exportSingleImage(jpegImagesFolderName: string, element: IAssetMetadata): Promise<void> {
```
Comment (wbreza) [inline src/providers/export/tensorFlowPascalVOC.ts]: This method is pretty large - I recommend you break this down into smaller easier to digest functions.

File: src/providers/export/tensorFlowPascalVOC.ts [hunk 3]
```diff
@@ -155,16 +197,147 @@ item {
         const annotationsFolderName = `${exportFolderName}/Annotations`;
         await this.storageProvider.createContainer(annotationsFolderName);
 
+        const annotationTemplate = `
+<annotation verified="yes">
```
Comment (wbreza) [inline src/providers/export/tensorFlowPascalVOC.ts]: Please move out to external file and see if we can import it like we do with JSON files.

File: src/providers/export/tensorFlowPascalVOC.ts [hunk 4]
```diff
@@ -155,16 +197,147 @@ item {
...
         // Save Annotations
-        // TODO
+        this.imagesInfo.forEach((imageInfo, imageName) => {
+            allAnnotationExports.push(
+                new Promise(async (resolve, reject) => {
+                    const imageFilePath = `${annotationsFolderName}/${imageName}`;
+                    const assetFilePath = `${imageFilePath.substr(0, imageFilePath.lastIndexOf("."))
+                        || imageFilePath}.xml`;
+
+                    const objectsXML = imageInfo.objects.map((o) => {
```
Comment (wbreza) [inline src/providers/export/tensorFlowPascalVOC.ts]: If you convert the `%variable% notation to use javascript `${variable}` notation you can then use the same string interpolation function that we use in localization.

File: src/providers/export/tensorFlowPascalVOC.ts [hunk 5]
```diff
@@ -155,16 +197,147 @@ item {
...
+
+        try {
+            await Promise.all(allAnnotationExports);
+        } catch (err) {
+            console.log(err);
+        }
     }
 
-    private async exportImageSets(exportFolderName: string, allAssets: IAssetMetadata[]) {
+    private async exportImageSets(exportFolderName: string, allAssets: IAssetMetadata[],
```
Comment (wbreza) [inline src/providers/export/tensorFlowPascalVOC.ts]: Split out into smaller, easier to understand methods.

File: src/providers/export/tensorFlowPascalVOC.ts [hunk 6]
```diff
@@ -155,16 +197,147 @@ item {
...
+
+                } else {
+                    const imageSetFileName = `${imageSetsMainFolderName}/${tag.name}.txt`;
+                    await this.storageProvider.writeText(imageSetFileName, tagsDict.get(tag.name).join("\n"));
+                }
+            });
+        }
+    }
+
+    private shuffle(a: any[]) {
```
Comment (wbreza) [inline src/providers/export/tensorFlowPascalVOC.ts]: What does this do and how is it used?

File: src/react/components/pages/export/exportPage.tsx [hunk 7]
```diff
@@ -74,6 +74,7 @@ export default class ExportPage extends React.Component<IExportPageProps> {
         };
 
         await this.props.actions.saveProject(projectToUpdate);
+        await this.props.actions.exportProject(projectToUpdate);
```
Comment (wbreza) [inline src/react/components/pages/export/exportPage.tsx]: FYI - We don't want this here right now.  It is also causing your CI test failure.  You can run export from the editor page by clicking the "Export" button in the toolbar.

File: src/providers/export/tensorFlowPascalVOC.ts (hunk 6 above)
Comment (JacopoMangiavacchi) [reply src/providers/export/tensorFlowPascalVOC.ts]: shuffle() it's a utility function to shuffle a generic array.  I call this method to shuffle each tag arrays before splitting in test and validation set.  In ML if you're not working with sequenced datapoint like for example time series it's a good practice to shuffle an input dataset before splitting it in test and training dataset.

Anyway I'll put some comment in the code
